Introduce RepoContext

edsko opened the pull request:
**NOTE**: This PR consists of two commits, not four; the first two commits listed here are actually #3029.

The first of these commits just moves `GlobalFlags` to its own module, nothing else; this avoids circular module dependencies in the next commit.

The next commit introduces `RepoContext`:

``` haskell
-- | Access to repositories
data RepoContext = RepoContext {
    -- | All user-specified repositories
    repoContextRepos :: [Repo]

    -- | Get the HTTP transport
    --
    -- The transport will be initialized on the first call to this function
  , repoContextGetTransport :: IO HttpTransport

    -- | Should we ignore expiry times (when checking security)?
  , repoContextIgnoreExpiry :: Bool
  }
```

The `RepoContext` records the information about all available repositories, and how to access them (that is, the transport). The rest of the commit is mostly boring stuff:

* All functions that previously took a `[Repo]` argument now take a `RepoContext` argument
* All functions that previously had a `HttpTransport` argument now take a `RepoContext` argument
* A handful of functions that had neither now also have this `RepoContext` argument 
* The definition of `IndexFile` is now

    ``` haskell
    data Index = RepoIndex RepoContext Repo | SandboxIndex FilePath
    ```
* Calls to `configureTransport` are replaced with calls to `repoContextGetTransport`. 

The `HttpTransport` returned by `repoContextGetTransport` is initialized on first use, and cached thereafter. This means that we now always use a single transport, and it gets initialized with the information from the `GlobalFlags` (previously it was initialized in a few places were we had lost track of the `GlobalFlags` and so had to provide `Nothing` for the second argument to `configureTransport`).

The `RepoContext` is important for two reasons:

* `cabal-install` was creating `Repos` upfront, but delaying the construction of the transport until needed. However, when constructing a secure repo, we need to know the transport up front. By coupling these two we fix the impedance mismatch; the security integration will refer to the transport created in the repo context to create the `HttpLib` instance (in other words, we will use the `HttpTransport` in the security integration).
* In the security integration as it stood, `Repo` was extended with a third constructor for secure repos, but this constructed carried a stateful `Repository`. This was causing difficulties for the `nix-local-build` branch, which needs the `Repo` type to remain serializable. In the new integration (not yet pushed) the `RepoSecure` constructor still exists, but no longer carries the actual stateful `Repository`; this mapping is available separately from the `RepoContext`.

ezyang commented:
I haven't read the patches. The general gist is something I can empathize a lot with: one of the most annoying things about reading and refactoring Cabal code is that there are just too many damn arguments. Refactoring so that we have data structures to carry these along helps. But I will warn that this DOES increase coupling (it's very hard to refactor one of these contexts into separate ones after the fact... just ask GHC) so I feel like it's important to make sure we "get it right." (I would even say that we should have a ReaderT IO monad for plumbing these contexts around but that's not for this PR...)

Looking at the summary: why can't we just eagerly initialize the transport? It's just a record with a bunch of functions in it... Having an IO in the record is a huge smell.

edsko commented:
@dcoutts was quite adamant we did not want to initialize the transport eagerly. Why is having something of type `getMeTheSharedX_InitializeIfYouHaveTo :: IO X` a code smell though? Seems reasonable enough to me.

edsko commented:
(If we wanted to decouple the transport from the `RepoContext`, and yet don't want to initialize the transport eagerly, we'd have to pass an additional argument of type `IO HttpTransport` to lots of functions. At it stood lots of functions weren't passed the transport at it because they created it locally.)

edsko commented:
Or alternatively remove the transport from the `RepoContext`, and have functions initialize the transport themselves as before, and keep this on-demand initialization of the transport as a private feature of the `RepoContext`, that would be another possibility. (This would mean that some functions would no longer need the `RepoContext`, though most still will.)

ezyang commented:
> @dcoutts was quite adamant we did not want to initialize the transport eagerly.

Then we need a comment somewhere saying why.

I think of it as a code smell because the type is too big: IO allows a lot more than the initialize and cache! And then I have to worry about whether or not the cache is thread-safe in case we are going to parallelize it (what happens if two people race to initialize the cache; do I end up with two transports? Looks like it, looking at the code.) It might almost even be better to use `unsafePerformIO` (which is synchronized correctly).

Code request: please use a variable name that is not `ctxt`. It's a repository context, not a general purpose context! How about `repo_ctx` (or `repoContext` which is used in some places already)?

ezyang commented:
O/W, LGTM.

23Skidoo commented:
LGTM, but I agree with @ezyang than on-demand initialisation needs a better rationale than "@dcoutts told us so".

edsko commented:
@23Skidoo really? That's generally all the rationale I need :stuck_out_tongue:

edsko commented:
Rebased on `master`.

dcoutts commented:
Re initialising transports on demand...

In the nix-local-build branch it's kind of important to avoid the cost of initialising the transport when we don't need it. In the new branch when you 'build', most of the time we don't need a transport. But occasionally you do (e.g. first build). So we don't want to initialise the transport every time we call 'build' as that impacts on the latency when nothing has changed (we go to quite a bit of effort with caching to make this so).

The other vital thing for the nix-local-build branch is that `Repo` remain serialisable, which is why we can't just stick the secure repo interface into that. We need to be able to serialise the result of the solver, and that contains each package's source `Repo`.

edsko commented:
@23Skidoo , @ezyang I've made three changes to the PR:

* The variable for the `RepoContext` is now called `repoContext` throughout.
* Switched from an `MVar` to an `IORef` to that `repoContextGetTransport` is thread-safe.
* Added a comment for `repoContextGetTransport` explaining why it is important that the transport does not get initialized eagerly:

``` haskell
    -- | Get the HTTP transport
    --
    -- The transport will be initialized on the first call to this function.
    --
    -- NOTE: It is important that we don't eagerly initialize the transport.
    -- Initializing the transport is not free, and especially in contexts where
    -- we don't know a-priori whether or not we need the transport (for instance
    -- when using cabal in "nix mode") incurring the overhead of transport
    -- initialization on _every_ invocation (eg @cabal build@) is undesirable.
  , repoContextGetTransport :: IO HttpTransport
```

I hope that addresses all concerns. Let me know if this is now good to merge.

edsko commented:
Darn, I missed a few `ctxt`s. Searching for `\<ctxt\>` in git log -p does not work :)

edsko commented:
... and Atom's "search whole words only" is also not very smart :rage:

edsko commented:
Ok, they're really gone now (I think :flushed:)

ezyang commented:
OK, fine, LGTM.

(One day, I will convince people to use snake-case for local identifiers, so they are less likely to conflict with global camelcase identifiers...(

23Skidoo commented:
> Let me know if this is now good to merge.

Yes, should be good to go with these changes. Thanks!

dcoutts commented:
Ok then...

edsko commented:
@BardurArantsson Indeed, the `HttpLib` abstraction in `hackage-security` is much more bracket-like, and the mismatch between the two is causing some (minor) problems right now; there is still an open ticket about this in the hackage-security project https://github.com/well-typed/hackage-security/issues/42.
